[MRG] Add celltypes to `CellResponse` interface and `Network.rename_cell()`

asoplata opened the pull request:
This PR was originally much more complex, but has been greatly simplified by moving the discussion of hardcoded-celltypes to #972. This PR has now been changed to simply be an attempt to merge in the parts of #702 that can be made to pass our tests.

asoplata commented:
It is known that the tests are failing "across the board" after merging in #702 's prior work. These are mostly tied to new `cell_types` arguments and argument order for `CellResponse` and `Network`. I should be able to fix these issues as of tomorrow.

asoplata commented:
As I've been going through this and trying to get it back to passing all tests, I've noticed a minor "bug" so to speak, that `CellResponse`'s handling of its `cell_types` is inconsistent, and does not only include "cell types". `cell_response.py::CellResponse.cell_types()` returns a list of the unique labels of `self.spike_types`, but judging from both the docstring for that attribute and some test code, that list of labels includes both "cell types" which have spiked *and drives* which have spiked. Here is some example code that illustrates this:

```{python}
#!/usr/bin/env python

from hnn_core import jones_2009_model, MPIBackend, simulate_dipole
net = jones_2009_model()
location = 'proximal'
burst_std = 20
weights_ampa_p = {'L2_pyramidal': 1.0, 'L5_pyramidal': 1.0}
syn_delays_p = {'L2_pyramidal': 0.1, 'L5_pyramidal': 1.}

net.add_bursty_drive(
    'alpha_prox', tstart=50., burst_rate=10, burst_std=burst_std, numspikes=2,
    spike_isi=10, n_drive_cells=10, location=location,
    weights_ampa=weights_ampa_p, synaptic_delays=syn_delays_p, event_seed=284)

# When I run this longer simulation, I get the following, which includes the drive:
# net.cell_response.cell_types are ['L2_basket', 'L2_pyramidal', 'L5_basket', 'L5_pyramidal', 'alpha_prox']

with MPIBackend(n_procs=8):
    dpl = simulate_dipole(net, tstop=310., n_trials=1)

print(f'net.cell_response.cell_types are {net.cell_response.cell_types}')

# When I run this shorter simulation, I get the following, which is missing 'L5_basket':
# net.cell_response.cell_types are ['L2_basket', 'L2_pyramidal', 'L5_pyramidal', 'alpha_prox']

with MPIBackend(n_procs=8):
    dpl = simulate_dipole(net, tstop=30., n_trials=1)

print(f'net.cell_response.cell_types are {net.cell_response.cell_types}')
```

This means that `CellResponse.cell_types` cannot be relied on for "cell type" information, as opposed to "unique spike types" which is a better name for it (similar to how it's used here https://github.com/jonescompneurolab/hnn-core/blob/master/hnn_core/viz.py#L608 ).

I see two ways to approach this:

1. I think is the best way: We remove all "cell type" language from `CellResponse`'s use, and instead treat all "unique spike types that have spiked during the trial" the same.
2. We split the spike data into two parts, one for spikes of true "cell types" and one for spikes of drives. This is very clunky and bad design, including that eventually, we may want to add other things besides celltypes and drives.

Regardless, we should make discriminating between "cell type" spikes and "drive" spikes very clear and easy, including in plotting. Just like with wagdy88's changed, we could force the `CellResponse` user to supply true cell type names, and then we would always know if a spike type is a celltype or a drive. We could then (optionally) have a default true flag for plotting functions that only plots true celltypes by default (as opposed to drives). This would also make other things easier, like informing the user who is plotting a celltype that, if it is not found but in `_cell_type_names`, then that celltype can't be plotted because it produced no spikes, not that it's an incorrect name, etc. We could also remove all the celltype defaults from the plotting code that `CellResponse` uses, and also have a mapping at initialization time between all unique spike types and colors, so the colors are consistent.

rythorpe commented:
I agree, `CellResponse`'s handling of `cell_types` of very confusing. I like solution (1) where you convert everything to "`spike_types`" instead, which when compared internally to `_cell_type_names` can be used to delineate between cell vs. drive spikes.

jasmainak commented:
@asoplata I'll go through this PR ... just wanted to mention that you should avoid making branches on `upstream` as much as possible. It keeps the upstream repo clean. The only branches you want to be making are for each of the releases. Anyone with admin writes can push to any PR, so you can just start with a normal PR on your fork and anyone with push access to hnn-core can also push to the branch on your fork

jasmainak commented:
I agree `cell_types` are somewhat of a misnomer. But just be wary of changing things that may already be used in other people's codes. If it's really inconvenient, make sure to deprecate it ... I can help you organize a proper deprecation warning if you have never done it before.

For this PR, I would recommend creating a plan which breaks down the contribution into smaller PRs each of which are easy to review and merge. My initial thinking was to make `net._add_cell_type` public, then see how much of the pipeline works for a new cell type ... e.g., does the plotting of spikes work etc. Once you have that work, then in a separate PR you can implement more advanced features

jasmainak commented on hnn_core/network.py:
```diff
@@ -448,20 +450,23 @@ def __init__(self, params, add_drives_from_params=False,
-                 len(self.pos_dict['L5_basket'])))
-        return '<%s | %s>' % (class_name, s)
+        # Dynamically create the description based on the current cell types
+        descriptions = []
```
```suggestion
        descriptions = list()
```

just a stylistic suggestion, it's hard to distinguish `[]` and `{}` but `list()` and `dict()` are fairly clear and unmistakable.

asoplata commented:
> @asoplata I'll go through this PR ... just wanted to mention that you should avoid making branches on `upstream` as much as possible. It keeps the upstream repo clean. The only branches you want to be making are for each of the releases. Anyone with admin writes can push to any PR, so you can just start with a normal PR on your fork and anyone with push access to hnn-core can also push to the branch on your fork

@jasmainak Yeah, I'm not sure why I thought this approach was the best. Apologies! Originally, I thought that for a potentially very-large feature with lots of development on it, it would be good to make a "feature branch" out of #702, and then use this PR as the main place for organizing our hardcoding-removal efforts; other, smaller PRs could go into this feature-branch before the whole branch goes to `master`. The more I thought about it after your comment, however, the more I realized the disadvantages, including:

1. devs would have to do things in a nonstandard way (at all)
2. devs would have to target their own PR-branches to a different branch that is *not necessarily* in-sync with master, `master<->feature<->individualPR`,
3. devs would have to build and rebase onto a potentially-volatile feature-branch instead of `master`. In contrast, I would argue that helping people rebase their own changes on top of `master` locally is, itself, at least medium-complexity. Adding more complexity to that is a bad idea.
4. Even if is difficult to break down a major feature into smaller parts, doing so makes `master`'s history significantly easier to comprehend (i.e. the same reason we use "linear history" in the first place), and allows better and independent analysis of each component of said feature (each individual PR). It also means there is just "one workspace context" so to speak.
5. Intermingling of the actual code for the PR versus discussion (especially about things not located specifically in this PR) would make the conversation harder to read.
6. etc.

Instead, I should have fixed and merged #702, then made an Issue unbound to specific PRs. As such, I will move the discussion points over to a new issue #972, and then we'll go about it the usual way: many smaller PRs into `master`.

I will make one more small commit to remove a committed TODO related specifically to this PR number, then this PR should *actually* be merged.

rythorpe commented on hnn_core/tests/test_network.py:
```diff
@@ -1158,3 +1163,37 @@ def test_only_drives_specified(self, base_network, src_gids,
+    rename_pairs = list(zip(net.cell_types.keys(), new_names))
+    for original_name, new_name in rename_pairs:
+        net.rename_cell(original_name, new_name)
+    for new_name in new_names:
+        assert new_name in net.cell_types.keys()
+        assert new_name in net.pos_dict.keys()
```
We should also test that `original_name` isn't a key anymore

asoplata replied:
Done in latest commit.

jasmainak commented:
Sorry I haven't had the time to look yet! Please feel free to ping me again on the relevant PRs if I don't respond back in a few days.

asoplata commented:
No rush, @jasmainak ; just so you know, I had to force-push solely due to a conflict in `whats_new`. Currently, the only differences between the version you reviewed 2 weeks ago are that `whats_new` change, and the additional test suggestion by Ryan.

jasmainak commented on hnn_core/tests/test_network.py:
```diff
@@ -1158,3 +1163,39 @@ def test_only_drives_specified(self, base_network, src_gids,
+    # Tests for non-string new_name
+    new_name = 5
+    with pytest.raises(TypeError, match="new_name must be an instance of str"):
+        net.rename_cell('L2_basket_test', 5)
```
can you add some tests to check if plotting spike histograms and such works with the renamed cell types?

asoplata replied:
I will try to add a test using `plot_spikes_raster`, since it appears to be the most amenable to using different names for the celltypes. However, many of the plotting functions make assumptions about the cell type names that could possibly be broken by the renamed types (including `plot_spikes_hist`). 

To be verbose, there is a big difference between "adding functionality to rename celltypes in `Network`" and making sure that the rest of `hnn-core` can *successfully use* celltypes which have been renamed. I see this PR as adding the former, but for the latter, it is a larger scope of work (see #972). I see two ways we could approach this:

1. We merge this PR and add the celltype rename functionality sooner rather than later. This lets a user rename celltypes in `Network` but *does not guarantee* that that new `Network` will successfully work with the entire API, until #972 is finished. This PR would be one of the very-many smaller PRs needed to close #972.
2. Alternatively, we wait to merge this PR until after #972 is mostly complete, and the API is mature with respect to variable cell type names.

Which do you think @jasmainak ?

jasmainak commented on hnn_core/network.py:
```diff
@@ -1196,12 +1198,46 @@ def _add_cell_type(self, cell_name, pos, cell_template=None):
             self._n_cells += len(pos)
 
+    def rename_cell(self, original_name, new_name):
+        """Renames cells in the network and clears connectivity so user can
+        set new connections.
```
should it clear connectivity? That seems like an odd behavior ... I would have expected to keep the original connectivity

asoplata replied:
It appears to clear behavior due to the difficulty of changing celltype names in-place for connectivity and drive data. I will work on adding this

jasmainak commented:
@asoplata the way I envision this to work is the following. You can do something like this:

```py
net1 = jones_2009_model()
net2 = net.copy()
net1.rename_cells({f'old_name_col1': old_name for old_name in net1.cell_names})
net2.rename_cells({f'old_name_col2': old_name for old_name in net2.cell_names})

# now add the connections between the two networks -- separate PR
net = net1 + net2
net.add_connection('L5_pyramidal_col1', 'L5_pyramidal_col2', ...)
```

asoplata commented:
Somewhat unrelated note: Adding support for multi-network simulation (including distinct long-range connections between networks) is one of our GSoC 2025 project proposals: https://github.com/jonescompneurolab/hnn-core/wiki/GSoC-Ideas-2025

asoplata commented:
Had to force-push due to rebase conflict with `master` (the Sphinx theme change). I believe I have addressed the remaining comments in the most recent 2 commits, and barring future comments, this should be ready for merge.

Re: plotting tests, see my response above. I added several more tests that should satisfy overall. I was able to get `plot_spikes_raster` working with custom cell types, but only when forcing the network to have spikes: it only allows for custom celltype names if there are spikes present by those names in `CellResponse` (see #972). As an example, here is a live result with new names:

![image](https://github.com/user-attachments/assets/0ae7fdfb-91f2-4393-9e8e-9375bb1e0151)

Re: connectivity-clearing, I made major improvements to the newly-renamed `Network.rename_cell_type()`. It no longer clears the connectivity, but instead it now dives into all common attributes that make use of custom cell_type names that I could find, including connectivity, drives, and biases, and then updates the appropriate cell type names. There may be other uses of cell type names in the ultimate `Network` data, but I could not find any across multiple different networks with all the variations I could think of. The only exception is the `Cell.name` children of `Network`, but these were not updated for two reasons: 1. `Cell.name` currently has no way to be updated after initialization, and 2. IIRC these are important for actual NEURON cell building, and any renaming needs to be done carefully in league with #972.

asoplata commented:
Note that I believe the current CI failures are *not* due to the code in this actual PR, but instead something else breaking with our CI system. As an example, https://github.com/asoplata/hnn-core/actions/runs/13209404408/job/36879899559#step:10:1235 this shows that our MPI installation is no longer working in our unit tests. Will investigate

asoplata commented:
Had to rebase in order to include CI hotfix (#994) needed to pass tests.

rythorpe commented on doc/whats_new.md:
```diff
@@ -9,8 +9,13 @@ orphan: true
 
 ### Changelog
 
+<<<<<<< HEAD
```
Looks like the rebase was never completed.

asoplata replied:
Good catch! Thanks for finding that. I'm surprised my software let me continue the rebase as if it was done. Just fixed this.

jasmainak commented on hnn_core/tests/test_network.py:
```diff
@@ -1158,3 +1164,90 @@ def test_only_drives_specified(self, base_network, src_gids,
+    plot_dipole(dpls2, show=False)
+
+    dpls3 = simulate_dipole(net2, tstop=100., n_trials=1)
+    plot_dipole(dpls3, show=False)
```
what about the `cell_response` plots? Do they work too?

asoplata replied:
Added tests for them in latest commits.

jasmainak commented on hnn_core/network.py:
```diff
@@ -1196,12 +1198,81 @@ def _add_cell_type(self, cell_name, pos, cell_template=None):
             self.cell_types.update({cell_name: cell_template})
             self._n_cells += len(pos)
 
+    def rename_cell_type(self, original_name, new_name):
```
Thinking about it a bit, it might be convenient to have a wrapper function for `rename_cell_type` called `rename_cells` and it could simply accept a dictionary with key/value pairs that have the old name and new name. That might be the easiest for users.

Somewhat similar to: https://mne.tools/stable/generated/mne.io.Raw.html#mne.io.Raw.rename_channels

asoplata replied:
Hmm a dictionary approach is a great idea. I'm more inclined to, instead of using a wrapper to provide an additional interface, force a dictionary mapping input to be the only interface. In other words, remove the `original_name, new_name` arguments entirely and replace with dict like you suggested in a later comment in the tests. This would 1. stick to the "there should be one and only one way to do things" guideline, 2. would ease the maintenance burden by obviating any wrappers or having to handle multiple interfaces, and 3. still be just as easy as the singleton case, but also makes it trivial to provide multiple sets of inputs.

asoplata replied:
I've completed the switch to an exclusively-dictionary argument in the most recent commits, so this is probably resolved.

jasmainak commented on hnn_core/network.py:
```diff
@@ -1196,12 +1198,81 @@ def _add_cell_type(self, cell_name, pos, cell_template=None):
+            for bias_key, bias_value in self.external_biases.items():
+                if original_name in self.external_biases[bias_key].keys():
+                    self.external_biases[bias_key][new_name] = \
+                        self.external_biases[bias_key].pop(original_name)
+            # Update Network.external_drives
```
```suggestion
                        self.external_biases[bias_key].pop(original_name)
            
            # Update Network.external_drives
```

code needs a bit of breathing space ... it's like paragraphs in a paper :)

asoplata replied:
Done where you suggested and elsewhere.

jasmainak commented on hnn_core/network.py:
```diff
@@ -1196,12 +1198,81 @@ def _add_cell_type(self, cell_name, pos, cell_template=None):
+                    drive_config['target_types'].append(new_name)
+                    drive_config['target_types'].sort()
+                    for config_key, config_value in drive_config.items():
+                        if (config_key == 'dynamics'):
```
```suggestion
                        if config_key == 'dynamics':
```

just a stylistic suggestion

jasmainak commented on hnn_core/network.py:
```diff
@@ -1196,12 +1198,81 @@ def _add_cell_type(self, cell_name, pos, cell_template=None):
+            # Update Network.external_drives
+            for drive_key, drive_config in self.external_drives.items():
+                if original_name in drive_config['target_types']:
+                    drive_config['target_types'].remove(original_name)
+                    drive_config['target_types'].append(new_name)
+                    drive_config['target_types'].sort()
```
this part of the code looks a bit like it could be susceptible to breaking ... just a hunch. I wonder if it might be better to just construct a new dictionary with the correct keys -- if it's easy enough. Maybe you can ask copilot for help? :)

jasmainak replied:
typically if you try to modify an iterable while iterating over it, it can cause unforeseen problems. For example, if you remove certain elements from a list during the iteration, your loop may exit prematurely. I think it's a bit "unsafe" in certain scenarios. You might be fine here, but worth thinking about future contributors etc. who may not be completely comfortable in Python.

asoplata replied:
Okay, I looked into it a little more, and I'm going with your suggestion to not modify the dict being iterated over in-place. If I'm reading the `iter(dictview)` of https://docs.python.org/3/library/stdtypes.html#dictionary-view-objects correctly, modifying (adding/deleting) the keys of the dict will get us into trouble. This is similar to adding/deleting elements of a list you're iterating over, like you say. *Technically* I think this current code doesn't run into this trap because it is only changing values of the iterated-dict, not keys, but I will change it since this is bad practice. Some of the reasons why I thought the current code may be fine under modern python are that 1. it no longer complains if you change the size of the dict by modifying values of the iterated-dict like it used to, and 2. all dict entries are ordered now. I see now that the issue you brought up still applies.

jasmainak commented on hnn_core/network.py:
```diff
@@ -1196,12 +1198,81 @@ def _add_cell_type(self, cell_name, pos, cell_template=None):
+                                    config_value['rate_constant'].pop(
+                                        original_name
+                                    )
+                        elif ((config_key == 'synaptic_delays')
```
```suggestion
                        elif (config_key == 'synaptic_delays'
```

jasmainak replied:
I agree that there are certain cases in which the paranthesis does help. For example, in complex conditionals or multi-line conditionals, it is preferable to use paranthesis instead of backslash (\). This particular one didn't seem like that to me, but you can decide :)

jasmainak commented on hnn_core/network.py:
```diff
@@ -1196,12 +1198,81 @@ def _add_cell_type(self, cell_name, pos, cell_template=None):
+                             and (isinstance(config_value, dict))):
+                            drive_config[config_key][new_name] = \
+                                drive_config[config_key].pop(original_name)
+                        elif ('weights_' in config_key):
```
```suggestion
                        elif 'weights_' in config_key:
```

jasmainak commented on hnn_core/network.py:
```diff
@@ -1196,12 +1198,81 @@ def _add_cell_type(self, cell_name, pos, cell_template=None):
+
+            # Update Network.gid_ranges: order matters for consistency!
+            for _ in range(len(self.gid_ranges)):
+                name, gid_range = self.gid_ranges.popitem(last=False)
```
same comment as above ... I wonder if it might be easier/simpler to construct a new dictionary

asoplata replied:
See prior comment about new dict construction.

asoplata replied:
I'll also change this so it doesn't modify the dict in-place.

asoplata replied:
See above, this is probably resolved.

jasmainak commented on hnn_core/tests/test_cell_response.py:
```diff
@@ -57,11 +60,12 @@ def test_cell_response(tmp_path):
-    assert list(cell_response.__dict__.keys()) == \
-        sim_attributes + net_attributes
+    assert sorted(list(cell_response.__dict__.keys())) == \
+        sorted(sim_attributes + net_attributes)
```
the manual sorting scares me a bit

asoplata replied:
To be clear, the tests pass without this sorting present, so strictly speaking this change is unnecessary. However, without the sorting, in order for our tests to pass, I believe the order of the attributes in `sim_attributes + net_attributes` must match the *order* in which the attributes are *created* in `CellResponse.__init__()`. That seems very brittle to me, and I think that the only think that should matter is that the attributes are present, not their order. We can remove this if you want though.

jasmainak replied:
okay I think this was related to my concern about popping dictionary keys while iterating over it. I was worried that it changed the order of the dictionary elements in unexpected ways. I'll let you take the final call -- if you feel confident in this change, please let me know.

asoplata replied:
I will make changes in the prior comments so that the dict being iterated over is not changed in-place.

However, when it comes to "chang[ing] the order of the dictionary elements in unexpected ways", in general, I don't think it's a good idea to rely on the order of keys in a dictionary (as opposed to the order of a list), even if they are default ordered based on insertion. I'm curious what you think about this, since while I don't think it applies as much to `CellResponse`, there are probably other classes and dicts where their attributes/keys are going to be modified, and the order will be different from initialization. However, since there's already a way to access any key-value pair that you want regardless of its order, I don't think you should (in general) need to rely on the order of the key-value pairs. What are your thoughts?

jasmainak commented on hnn_core/tests/test_network.py:
```diff
@@ -1158,3 +1164,90 @@ def test_only_drives_specified(self, base_network, src_gids,
+    net2 = net1.copy()
+    assert net2.connectivity
+    # adding a list of new_names
+    new_names = ['L2_basket_test', 'L2_pyramidal_test',
+                 'L5_basket_test', 'L5_pyramidal_test']
+    # avoid iteration through net.cell_type.keys() by creating tuples of old and new names
+    rename_pairs = list(zip(net2.cell_types.keys(), new_names))
+    for original_name, new_name in rename_pairs:
+        net2.rename_cell_type(original_name, new_name)
```
yeah this could become simpler with the new function I was suggesting:

```py
mapping = {cell_name: f'{cell_name}_test for cell_name in net2.cell_types}
net.rename_cells(mapping)
```

jasmainak replied:
I am also in two minds whether to make this method public at all ... I would recommend keeping everything private until you have worked through your series of PRs. Once you make a method public, the users will fall in love with it whether or not it's really needed and it becomes hard to change the code and reduces your degrees of freedom to refactor.

It's worth thinking through what the final API could look like. Perhaps:

```py
net1 = jones_2009_model()
net2 = net1.copy()
prefixes = ['net1', 'net2']

net = add_networks(nets, prefixes)
```

so then the user doesn't even ever attempt to rename cells in a non-standard way ...

asoplata replied:
Re: passing a dictionary instead, see my prior comment; I will make changes so the main interface uses a dictionary.

asoplata replied:
Both of these concerns should have been resolved in the most recent commits; the interface was changed to dictionary-only, and func was made private.

jasmainak commented on hnn_core/tests/test_network.py:
```diff
@@ -1158,3 +1164,90 @@ def test_only_drives_specified(self, base_network, src_gids,
+    #
+    # Test that the networks actually run
+    #
+    dpls1 = simulate_dipole(net1, tstop=100., n_trials=1)
```
won't these tests be slow?

asoplata replied:
If you're referring to the length of the `tstop`, I will decrease those.

asoplata replied:
Done.

jasmainak commented on doc/whats_new.md:
```diff
@@ -12,6 +12,9 @@ orphan: true
   by [Dikshant Jha][] in {gh}`987`.
 
+- Remove hardcoding of celltypes in :class:`~hnn_core.CellResponse` and add
+  {func}`~hnn_core.Network.rename_cell`, by [Mohamed W. ElSayed][] in {gh}`702` and {gh}`970`.
```
you should also add your name @asoplata :)

asoplata replied:
meh, I think it's fine without, there will be plenty of future additions with my username in them heh.

jasmainak commented:
apologies @asoplata for the slow review! The code is beginning to take really good shape ...

asoplata commented:
Force-pushed a rebase onto recent `master` for #1010

asoplata commented:
Okay @jasmainak , I believe I've either incorporated (and resolved) all of your changes, or responded to them.

jasmainak commented on hnn_core/tests/test_network.py:
```diff
@@ -1158,3 +1164,92 @@ def test_only_drives_specified(self, base_network, src_gids,
+    net4 = hnn_core.hnn_io.read_network_configuration(
+        op.join(hnn_core_root, 'tests', 'assets', 'jones2009_3x3_drives.json'))
+    net4._rename_cell_types(cell_type_rename_mapping)
+    dpls4 = simulate_dipole(net4, tstop=2., n_trials=1)
+    plot_dipole(dpls4, show=False)
+    net4.cell_response.plot_spikes_raster(show=False)
+    net4.cell_response.plot_spikes_hist(show=False)
```
is this actually testing anything here? Since `tstop=2`, there will not be any spikes from the new cell type I presume?

asoplata replied:
I'll test how long it takes for a spike and then update the time lengths.

asoplata replied:
In the latest commits, I extended the sim time and confirmed there is spiking. I also added missing code for `plot_spikes_raster` that was required for plotting the non-canonical celltype names. This should be resolved now.

jasmainak commented:
LGTM besides a few last comments.

asoplata commented:
@jasmainak Have you had a chance to look at the final response comments? I think I addressed your concerns

jasmainak commented:
see #1045 !

asoplata commented on hnn_core/network.py:
```diff
@@ -447,12 +447,15 @@ def __init__(self, params, add_drives_from_params=False,
-              % (len(self.pos_dict['L2_basket']),
-                 len(self.pos_dict['L5_basket'])))
-        return '<%s | %s>' % (class_name, s)
+        # Dynamically create the description based on the current cell types
```
Added a test for `Network.__repr__()` in 2107bacdd81ecd92a836724cb9a4356dfba061bc .

jasmainak commented:
Thanks @asoplata for taking this to the finish line! 🥳 🎈
